### PR TITLE
feature(http): adds API for handling HTTP responses

### PR DESCRIPTION
**Pull request by mrclay**
Adds API for handling HTTP responses from pages and actions in a unified and testable manner.
Adds `Elgg\Http\ResponseFactory` service for preparing and serving responses.
Adds `Elgg\Http\ResponseBuilder` interface as well as concrete classes for common response types.

Adds `elgg_ok_response()`, `elgg_error_response()` and `elgg_redirect_response()`. These functions can be used to provide testable return from action files and page handlers.

Replaces #9805. Now targeting 2.3.
- [x] Add test for `_elgg_csrf_token_refresh()` response

### CODE REVIEW

File: engine/classes/Elgg/ActionsService.php [hunk 1]
```diff
@@ -334,6 +393,21 @@ public function generateActionToken($timestamp, $session_token = '') {
 		return _elgg_services()->crypto->getHmac([(int)$timestamp, $session_token], 'md5')
 			->getToken();
 	}
+
+	/**
```
Comment (mrclay) [inline engine/classes/Elgg/ActionsService.php]: No longer used

**Comment by mrclay** [comment]
@hypeJunction I think we can just delete branch `2.x_9779`, right?

**Comment by hypeJunction** [comment]
Right. We didn't merge anything in there

**Comment by hypeJunction** [comment]
@mrclay Do we need to update action token logic?

File: engine/lib/actions.php [hunk 2]
```diff
@@ -303,10 +319,8 @@ function _elgg_csrf_token_refresh() {
 		'user_guid' => elgg_get_logged_in_user_guid(),
 	);
 
-	header("Content-Type: application/json;charset=utf-8");
-	echo json_encode($data);
-
-	return true;
+	elgg_set_http_header("Content-Type: application/json;charset=utf-8");
+	return elgg_ok_response($data);
```
Comment (mrclay) [inline engine/lib/actions.php]: @hypeJunction here's all I changed for the token refresh. I just deleted your test (it didn't cover the new behavior), but I'll try to add the test back.

**Comment by hypeJunction** [comment]
I made a PR against your leaf moving the upgrading guides. Let's merge and then add the token test. I would like to start rolling out some other ideas that are flooding my brain.

**Comment by hypeJunction** [comment]
Let's just merge, I will update the docs afterwards. Will save us time

**Comment by mrclay** [comment]
Taking the plunge!

**Comment by hypeJunction** [comment]
I am confident this will be a big improvement once we are done. Having testabel routes and actions is a big leap forward. I will add HTTP exceptions next, that will speed up writing actions. One thing that we didn't consider here is webservices. Need to write some tests, but need a plugin testing strategy first.

**Pull request merged**